pkg/cluster/cluster.go

```diff
@@ -82,63 +83,86 @@ type Cluster struct {
 	backupDir string
 }
 
-func New(c Config, e *spec.EtcdCluster, stopC <-chan struct{}, wg *sync.WaitGroup) (*Cluster, error) {
-	return new(c, e, stopC, wg, true)
-}
+func New(config Config, e *spec.EtcdCluster, stopC <-chan struct{}, wg *sync.WaitGroup) *Cluster {
+	lg := logrus.WithField("pkg", "cluster").WithField("cluster-name", e.Name)
+	c := &Cluster{
+		logger:  lg,
+		config:  config,
+		cluster: e,
+		eventCh: make(chan *clusterEvent, 100),
+		stopCh:  make(chan struct{}),
+		status:  &spec.ClusterStatus{},
+	}
 
-func Restore(c Config, e *spec.EtcdCluster, stopC <-chan struct{}, wg *sync.WaitGroup) (*Cluster, error) {
-	return new(c, e, stopC, wg, false)
+	err := c.setup()
+	if err != nil {
+		c.logger.Errorf("fail to setup: %v", err)
+		if c.status.Phase != spec.ClusterPhaseFailed {
+			c.status.SetReason(err.Error())
+			c.status.SetPhase(spec.ClusterPhaseFailed)
+			if err := c.updateStatus(); err != nil {
+				c.logger.Errorf("fail to set failed phase: %v", err)
+			}
+		}
+		return nil
+	}
+
+	wg.Add(1)
+	go c.run(stopC, wg)
+	return c
 }
 
-func new(config Config, e *spec.EtcdCluster, stopC <-chan struct{}, wg *sync.WaitGroup, isNewCluster bool) (*Cluster, error) {
-	err := e.Spec.Validate()
+func (c *Cluster) setup() error {
+	err := c.cluster.Spec.Validate()
 	if err != nil {
-		return nil, fmt.Errorf("invalid cluster spec: %v", err)
+		return fmt.Errorf("invalid cluster spec: %v", err)
 	}
 
-	lg := logrus.WithField("pkg", "cluster").WithField("cluster-name", e.Name)
-	var bm *backupManager
+	var isNewCluster bool
+	switch c.status.Phase {
+	case "":
+		isNewCluster = true
+	case spec.ClusterPhaseCreating:
+		return errors.New("cluster creation was attempted before but not finished")
+	case spec.ClusterPhaseRunning:
+		isNewCluster = false
+	case spec.ClusterPhaseFailed:
+		return errors.New("cluster already failed")
+	}
 
-	if b := e.Spec.Backup; b != nil && b.MaxBackups > 0 {
-		bm, err = newBackupManager(config, e, lg, isNewCluster)
+	if b := c.cluster.Spec.Backup; b != nil && b.MaxBackups > 0 {
+		c.bm, err = newBackupManager(c.config, c.cluster, c.logger, isNewCluster)
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
-	c := &Cluster{
-		logger:  lg,
-		config:  config,
-		cluster: e,
-		eventCh: make(chan *clusterEvent, 100),
-		stopCh:  make(chan struct{}),
-		status:  &spec.ClusterStatus{},
-		bm:      bm,
-	}
 
 	if isNewCluster {
```

Comment:
we can remove this if statement. we can return early if we do not need the actual setup